www/settime

```diff
@@ -0,0 +1,132 @@
+<?php
+
+include_once "session-start.php";
+include_once "util.php";
+include_once "pagetpl.php";
+include_once "login-check.php";
+
+function sendResponse($statmsg, $errmsg, $detail)
+{
+    global $xml;
+
+    if ($xml)
+    {
+        header("Content-Type: text/xml");
+        header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
+        header("Cache-Control: no-store, no-cache, must-revalidate");
+
+        echo "<?xml version=\"1.0\" encoding=\"UTF-8\" ?><response>"
+            . ($statmsg ? "<label>$statmsg</label>" : "")
+            . ($errmsg ? "<error>$errmsg</error>" : "")
+            . "</response>";
+    }
+    else
+    {
+        smallPageHeader("Update Your Time Vote");   // Do we need this?
+        if ($errmsg)
+            echo "<span class=errmsg>$errmsg</span><p>";
+
+        echo $detail;
+        smallPageFooter();
+    }
+
+    exit();
+}
+
+// connect to the database
+include_once "dbconnect.php";
+$db = dbConnect();
+if ($db == false)
+    sendResponse("Not Saved", "An error occurred connecting to the database. "
+                 . "Please try again later.", false);
+
+// get the request parameters
+$game = get_req_data('game');
+$newtime = get_req_data('newtime');
+$xml = isset($_REQUEST['xml']);
+
+// make sure there's a game
+if ($game == "")
+    sendResponse("Not Saved", "No game was specified.", false);
+
+// make sure we process any persistent login state
+$userid = checkPersistentLogin();
+if (!$userid && !$xml && !logged_in(true))
+    exit();
+
+// make sure we're logged in
+if (!$userid)
+    sendResponse("Not Saved", "To submit a time, please log in.", false);
+
+// make sure the game is valid
+$qgame = mysql_real_escape_string($game, $db);
+$result = mysql_query("select id from games where id = '$qgame'", $db);
+if (mysql_num_rows($result) == 0)
+    sendResponse("Not Saved", "The specified game was not found.", false);
+
+// if there's an existing time for this user for this game, simply update
+// the time; otherwise add a new row
+
+$result = mysql_query("lock tables playertimes write", $db);
+if ($result) {
+    $result = mysql_query(
+        "select id, timevote from playertimes
+         where gameid='$qgame' and userid='$userid'", $db);
+}
+if ($result && mysql_num_rows($result)) {
```

Comment:
Maybe there's no need to check the current vote.

I think you can do the delete/update/insert based on the gameid and userid, instead of the row ID.

You can try the delete, and check the result to realize if it existed or not, and base the error/success message on that.

And insert/update can be done in a single "upsert", and use the result to find out which it was, to build the correct success message. (I don't know how to do that in PHP, but I assume there's a way)

Reply:
Almost everything on this page is copied from setrating, fyi. :)  I do not claim to understand it all.

(Maybe some things that were needed for setrating won't apply here? Setrating deals with both reviews and ratings, but here we are only dealing with times.)

Reply:
I'll submit a PR later converting it to the new JSON approach, and trying out my suggestion.